src/Microsoft.IdentityModel.Tokens/AsymmetricAdapter.cs

```diff
@@ -373,13 +385,45 @@ private byte[] SignUsingOffsetRsa(byte[] bytes, int offset, int count)
 
         private bool VerifyRsa(byte[] bytes, byte[] signature)
         {
+#if NET6_0_OR_GREATER
+            return VerifyUsingSpan(isRSA: true, bytes, signature);
+#else
             return RSA.VerifyHash(HashAlgorithm.ComputeHash(bytes), signature, HashAlgorithmName, RSASignaturePadding);
+#endif
         }
 
         private bool VerifyUsingOffsetRsa(byte[] bytes, int offset, int count, byte[] signature)
         {
+#if NET6_0_OR_GREATER
+            return VerifyUsingSpan(isRSA: true, bytes.AsSpan(offset, count), signature);
+#else
             return RSA.VerifyHash(HashAlgorithm.ComputeHash(bytes, offset, count), signature, HashAlgorithmName, RSASignaturePadding);
+#endif
+        }
+
+#if NET6_0_OR_GREATER
+        private bool VerifyUsingSpan(bool isRSA, ReadOnlySpan<byte> bytes, byte[] signature)
+        {
+            int hashByteLength = HashAlgorithm.HashSize / 8;
+            byte[] array = null;
+            Span<byte> hash = hashByteLength <= 256 ? stackalloc byte[256] : array = ArrayPool<byte>.Shared.Rent(hashByteLength);
+            hash = hash.Slice(0, hashByteLength);
+
+            bool hashResult = HashAlgorithm.TryComputeHash(bytes, hash, out int bytesWritten);
+            Debug.Assert(hashResult && bytesWritten == hashByteLength, "HashAlgorithm.TryComputeHash failed");
+
+            bool result = isRSA ?
```

Comment:
there are a few exceptions that could come from either of these methods, should we move the Shared.Return call to a try-finally?

Reply:
Done.